File: Example/iOS-Objective-C/ViewController.m
//
//  ViewController.m
//  iOS-Objective-C
//
//  Created by Felix Herrmann on 14.10.21.
//

#import "ViewController.h"
#import "iOS_Objective_C-Swift.h"
#import "PaintingViewController.h"
@import UIImageColorsObjc;


@interface UIPageViewController ()

@property UIPageControl *_pageControl; // expose internal property

@end


@interface ViewController () <UIPageViewControllerDataSource, UIPageViewControllerDelegate>

@property (strong) UILabel *mainLabel;
@property (strong) UILabel *secondaryLabel;
@property (strong) UILabel *detailLabel;

@property (strong) UIPageViewController *pageViewController;

@property (nonatomic) NSInteger currentIndex;

@end


@implementation ViewController


#pragma mark - Setter

- (void)setCurrentIndex:(NSInteger)currentIndex {
    if(currentIndex < 0) {
        _currentIndex = Painting.examples.count - 1;
    } else if (currentIndex > Painting.examples.count - 1) {
        _currentIndex = 0;
    } else {
        _currentIndex = currentIndex;
    }
}


#pragma mark - ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self setupLabels];
    [self setupPageViewController];
}


#pragma mark - Setup

- (void)setupLabels {
    self.mainLabel = [[UILabel alloc] init];
    self.mainLabel.translatesAutoresizingMaskIntoConstraints = false;
    self.mainLabel.numberOfLines = 2;
    self.mainLabel.font = [UIFont systemFontOfSize:25 weight:UIFontWeightRegular];
    
    self.secondaryLabel = [[UILabel alloc] init];
    self.secondaryLabel.translatesAutoresizingMaskIntoConstraints = false;
    self.secondaryLabel.font = [UIFont systemFontOfSize:20 weight:UIFontWeightRegular];
    
    self.detailLabel = [[UILabel alloc] init];
    self.detailLabel.translatesAutoresizingMaskIntoConstraints = false;
    self.detailLabel.font = [UIFont systemFontOfSize:15 weight:UIFontWeightRegular];
    
    [self.view addSubview:self.mainLabel];
    [self.view addSubview:self.secondaryLabel];
    [self.view addSubview:self.detailLabel];
    
    [NSLayoutConstraint activateConstraints:@[
        [self.mainLabel.leadingAnchor constraintEqualToAnchor:self.view.readableContentGuide.leadingAnchor constant:20],
        [self.mainLabel.topAnchor constraintEqualToAnchor:self.view.readableContentGuide.topAnchor constant:72],
        [self.mainLabel.trailingAnchor constraintLessThanOrEqualToAnchor:self.view.readableContentGuide.trailingAnchor constant:-20],
        
        [self.secondaryLabel.leadingAnchor constraintEqualToAnchor: self.view.readableContentGuide.leadingAnchor constant: 20],
        [self.secondaryLabel.trailingAnchor constraintLessThanOrEqualToAnchor: self.view.readableContentGuide.trailingAnchor constant: -20],
        [self.secondaryLabel.bottomAnchor constraintEqualToAnchor: self.detailLabel.topAnchor  constant: -4],
        
        [self.detailLabel.leadingAnchor constraintEqualToAnchor: self.view.readableContentGuide.leadingAnchor constant: 20],
        [self.detailLabel.trailingAnchor constraintLessThanOrEqualToAnchor: self.view.readableContentGuide.trailingAnchor constant: -20],
        [self.detailLabel.bottomAnchor constraintEqualToAnchor: self.view.readableContentGuide.bottomAnchor constant: -92]
    ]];
}

- (void)setupPageViewController {
    self.pageViewController = [[UIPageViewController alloc] initWithTransitionStyle:UIPageViewControllerTransitionStyleScroll navigationOrientation:UIPageViewControllerNavigationOrientationHorizontal options:nil];
    self.pageViewController.dataSource = self;
    self.pageViewController.delegate = self;
    
    PaintingViewController *firstPaitingViewController = [[PaintingViewController alloc] initWithPainting:Painting.examples[self.currentIndex]];
    [self.pageViewController setViewControllers:@[firstPaitingViewController] direction:UIPageViewControllerNavigationDirectionForward animated:false completion:nil];
    [self updateColorsFrom:firstPaitingViewController];
    
    [self addChildViewController:self.pageViewController];
    [self.pageViewController didMoveToParentViewController:self];
    [self.view addSubview:self.pageViewController.view];
    
    [NSLayoutConstraint activateConstraints:@[
        [self.pageViewController.view.leadingAnchor constraintEqualToAnchor: self.view.leadingAnchor],
        [self.pageViewController.view.topAnchor constraintEqualToAnchor: self.view.topAnchor],
        [self.pageViewController.view.trailingAnchor constraintEqualToAnchor: self.view.trailingAnchor],
        [self.pageViewController.view.bottomAnchor constraintEqualToAnchor: self.view.bottomAnchor]
    ]];
}


#pragma mark - Methods

- (void)updateColorsFrom:(PaintingViewController *)paintingViewController {
    [paintingViewController.painting.image getColorsWithQuality:UIImageColorsScaleQualityHigh completion:^(UIImageColors * _Nullable colors) {
        [UIView animateWithDuration:0.15 animations:^{
            self.view.backgroundColor = colors.background;
            
            self.mainLabel.text = paintingViewController.painting.name;
            self.mainLabel.textColor = colors.primary;
            
            self.secondaryLabel.text = paintingViewController.painting.artist;
            self.secondaryLabel.textColor = colors.secondary;
            
            self.detailLabel.text = [NSString stringWithFormat:@"%ld", paintingViewController.painting.year];
            self.detailLabel.textColor = colors.detail;
            
            self.pageViewController._pageControl.currentPageIndicatorTintColor = colors.primary;
            self.pageViewController._pageControl.pageIndicatorTintColor = colors.secondary;
        }];
    }];
}


#pragma mark - UIPageViewControllerDataSource

- (UIViewController *)pageViewController:(UIPageViewController *)pageViewController viewControllerBeforeViewController:(UIViewController *)viewController {
    PaintingViewController *paintingViewController = (PaintingViewController *)viewController;
    NSUInteger index = [Painting.examples indexOfObject:paintingViewController.painting];
    self.currentIndex = index - 1;
    return [[PaintingViewController alloc] initWithPainting:Painting.examples[self.currentIndex]];
}

- (UIViewController *)pageViewController:(UIPageViewController *)pageViewController viewControllerAfterViewController:(UIViewController *)viewController {
    PaintingViewController *paintingViewController = (PaintingViewController *)viewController;
    NSUInteger index = [Painting.examples indexOfObject:paintingViewController.painting];
    self.currentIndex = index + 1;
    return [[PaintingViewController alloc] initWithPainting:Painting.examples[self.currentIndex]];
}

- (NSInteger)presentationCountForPageViewController:(UIPageViewController *)pageViewController {
    return Painting.examples.count;
}

- (NSInteger)presentationIndexForPageViewController:(UIPageViewController *)pageViewController {
    return self.currentIndex;
}


#pragma mark - UIPageViewControllerDelegate

- (void)pageViewController:(UIPageViewController *)pageViewController didFinishAnimating:(BOOL)finished previousViewControllers:(NSArray<UIViewController *> *)previousViewControllers transitionCompleted:(BOOL)completed {
    PaintingViewController *paintingViewController = (PaintingViewController *)pageViewController.viewControllers.firstObject;
    [self updateColorsFrom:paintingViewController];
}

@end

File: Example/iOS/ViewController.swift
//
//  ViewController.swift
//  Example
//
//  Created by Felix Herrmann on 10.10.21.
//

import UIKit
import UIImageColors

final class ViewController: UIViewController {
    
    
    // MARK: - Properties
    
    private let mainLabel = UILabel()
    private let secondaryLabel = UILabel()
    private let detailLabel = UILabel()
    
    private let pageViewController = UIPageViewController(transitionStyle: .scroll, navigationOrientation: .horizontal)
    
    private var currentIndex: Int = 0 {
        didSet {
            if currentIndex < Painting.examples.startIndex {
                currentIndex = Painting.examples.endIndex - 1
            } else if currentIndex > Painting.examples.endIndex - 1 {
                currentIndex = Painting.examples.startIndex
            }
        }
    }
    
    
    // MARK: - ViewController
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        setupLabels()
        setupPageViewController()
    }
    
    
    // MARK: - Setup
    
    private func setupLabels() {
        mainLabel.translatesAutoresizingMaskIntoConstraints = false
        mainLabel.numberOfLines = 2
        mainLabel.font = .systemFont(ofSize: 25, weight: .regular)
        
        secondaryLabel.translatesAutoresizingMaskIntoConstraints = false
        secondaryLabel.font = .systemFont(ofSize: 20, weight: .regular)
        
        detailLabel.translatesAutoresizingMaskIntoConstraints = false
        detailLabel.font = .systemFont(ofSize: 15, weight: .regular)
        
        view.addSubview(mainLabel)
        view.addSubview(secondaryLabel)
        view.addSubview(detailLabel)
        
        NSLayoutConstraint.activate([
            mainLabel.leadingAnchor.constraint(equalTo: view.readableContentGuide.leadingAnchor, constant: 20),
            mainLabel.topAnchor.constraint(equalTo: view.readableContentGuide.topAnchor, constant: 72),
            mainLabel.trailingAnchor.constraint(lessThanOrEqualTo: view.readableContentGuide.trailingAnchor, constant: -20),
            
            secondaryLabel.leadingAnchor.constraint(equalTo: view.readableContentGuide.leadingAnchor, constant: 20),
            secondaryLabel.trailingAnchor.constraint(lessThanOrEqualTo: view.readableContentGuide.trailingAnchor, constant: -20),
            secondaryLabel.bottomAnchor.constraint(equalTo: detailLabel.topAnchor, constant: -4),
            
            detailLabel.leadingAnchor.constraint(equalTo: view.readableContentGuide.leadingAnchor, constant: 20),
            detailLabel.trailingAnchor.constraint(lessThanOrEqualTo: view.readableContentGuide.trailingAnchor, constant: -20),
            detailLabel.bottomAnchor.constraint(equalTo: view.readableContentGuide.bottomAnchor, constant: -92)
        ])
    }
    
    private func setupPageViewController() {
        pageViewController.dataSource = self
        pageViewController.delegate = self
        
        let firstPaintingViewController = PaintingViewController(painting: .examples[currentIndex])
        pageViewController.setViewControllers([firstPaintingViewController], direction: .forward, animated: false)
        updateColors(from: firstPaintingViewController)
        
        addChild(pageViewController)
        pageViewController.didMove(toParent: self)
        view.addSubview(pageViewController.view)
        
        NSLayoutConstraint.activate([
            pageViewController.view.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            pageViewController.view.topAnchor.constraint(equalTo: view.topAnchor),
            pageViewController.view.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            pageViewController.view.bottomAnchor.constraint(equalTo: view.bottomAnchor)
        ])
    }
    
    
    // MARK: - Methods
    
    private func updateColors(from paintingViewController: PaintingViewController) {
        paintingViewController.painting.image.getColors { colors in
            UIView.animate(withDuration: 0.15) { [self] in
                view.backgroundColor = colors?.background
                
                mainLabel.text = paintingViewController.painting.name
                mainLabel.textColor = colors?.primary
                
                secondaryLabel.text = paintingViewController.painting.artist
                secondaryLabel.textColor = colors?.secondary
                
                detailLabel.text = paintingViewController.painting.year.description
                detailLabel.textColor = colors?.detail
                
                pageViewController.pageControl?.currentPageIndicatorTintColor = colors?.primary
                pageViewController.pageControl?.pageIndicatorTintColor = colors?.secondary
            }
        }
    }
}


// MARK: - ViewController + UIPageViewControllerDataSource

extension ViewController: UIPageViewControllerDataSource {
    
    func pageViewController(_ pageViewController: UIPageViewController, viewControllerBefore viewController: UIViewController) -> UIViewController? {
        guard let paintingViewController = viewController as? PaintingViewController else { return nil }
        guard let index = Painting.examples.firstIndex(of: paintingViewController.painting) else { return nil }
        currentIndex = index - 1
        return PaintingViewController(painting: .examples[currentIndex])
    }
    
    func pageViewController(_ pageViewController: UIPageViewController, viewControllerAfter viewController: UIViewController) -> UIViewController? {
        guard let paintingViewController = viewController as? PaintingViewController else { return nil }
        guard let index = Painting.examples.firstIndex(of: paintingViewController.painting) else { return nil }
        currentIndex = index + 1
        return PaintingViewController(painting: .examples[currentIndex])
    }
    
    func presentationCount(for pageViewController: UIPageViewController) -> Int {
        return Painting.examples.count
    }
    
    func presentationIndex(for pageViewController: UIPageViewController) -> Int {
        currentIndex
    }
}


// MARK: - ViewController + UIPageViewControllerDelegate

extension ViewController: UIPageViewControllerDelegate {
    
    func pageViewController(_ pageViewController: UIPageViewController, didFinishAnimating finished: Bool, previousViewControllers: [UIViewController], transitionCompleted completed: Bool) {
        guard let paintingViewController = pageViewController.viewControllers?.first as? PaintingViewController else { return }
        updateColors(from: paintingViewController)
    }
}
